### DISCUSSION TITLE
Bind application to wireless network interface

### DISCUSSION THREAD

**Discussion by Aqua1ung** [netblue30/firejail#5525] [Q&A]
I need to bind ssmtp to my wireless network interface. Can this be done using firejail?

**Comment by netblue30** [comment]
You run it with --net option: firejail --net=ifname program-name

where ifname is the name of your wireless interface (wlan0 or whatever you have)

**Reply by Aqua1ung** [reply]
Thank you! I asked this because I seem to recall reading something on the firejail website about "only for regular Ethernet NICs" or something to that effect. Anyway, this is good news! Many thanks for your work!

**Reply by Aqua1ung** [reply]
ssmtp works just fine being launched by itself, but when I launch it using firejail:

`sudo firejail --net=wlp1s0  ssmtp -v myemail@email.com < ~/Documents/fille.txt`,

I get this error:

`Child process initialized in 1225.96 ms
ssmtp: Cannot open smtp.comcast.net:465`

What could be happening? Why does the server not like it?

Again, `ssmtp -v myemail@email.com < ~/Documents/fille.txt` sends the email with no trouble whatsoever.

**Reply by Aqua1ung** [reply]
[Here are some further investigations I conducted](https://www.reddit.com/r/linuxquestions/comments/znxt3d/comment/j0l87vj/?utm_source=reddit&utm_medium=web2x&context=3): I noticed a "security" email issued by my SMTP server to the effect that "a password is required," hence I am wondering if by firejailing ssmtp the password is being stripped out, hence the server rejects the connection. Could that be it? And yes, I am running firejail with the `--noprofile` option.

**Reply by ghost** [reply]
> sudo firejail --net=wlp1s0 ssmtp -v myemail@email.com < ~/Documents/fille.txt

When you want to have a visual check of the files being included by Firejail, add `--ignore=quiet`. Using `--noprofile` doesn't offer any protection (as the name suggests). More of a debugging tool really.

> Child process initialized in 1225.96 ms ssmtp: Cannot open smtp.comcast.net:465

Not sure what's exactly causing this error. But in my tests I've seen it too. I also happen to use ssmtp on a few machines and your thread inspired me to write a dedicated ssmtp.profile for it. Thanks for that! I never actually considered firejailing ssmtp before. Running ssmtp with those profile files the error is gone and sandboxing is more restrictive, pretty tight in fact.

You can find the relevant files [here](https://gist.github.com/glitsj16/76e6c6039e5d6ac4234b69d4686304ea). As usual (cfr. firecfg), if you want to be able to simply use `ssmtp` on the command-line and have it sandboxed automatically, create a symlink to /usr/bin/firejail in /usr/local/bin/ssmtp. As it happens, ssmtp provides the `sendmail` command too, so if you put a symlink to /usr/local/bin/ssmtp at /usr/local/bin/sendmail you can use both interchangeably and have them sandboxed.

If you could report back here with your observations, I could consider adding them to our collection. But let's wait a bit before opening a PR just yet.

**Reply by Aqua1ung** [reply]
> You can find the relevant files [here](https://gist.github.com/glitsj16/76e6c6039e5d6ac4234b69d4686304ea). As usual (cfr. firecfg), if you want to be able to simply use `ssmtp` on the command-line and have it sandboxed automatically, create a symlink to /usr/bin/firejail in /usr/local/bin/ssmtp. As it happens, ssmtp provides the `sendmail` command too, so if you put a symlink to /usr/local/bin/ssmtp at /usr/local/bin/sendmail you can use both interchangeably and have them sandboxed.
> 
> If you could report back here with your observations, I could consider adding them to our collection. But let's wait a bit before opening a > PR just yet.

Thank you! Security is not my main concern in running ssmtp firejailed. All I want is to make ssmtp use one _specific_ NIC and that is all. Nevertheless, if you say installing this profile should make it work, then I am certainly willing to give it a try. Now I have to learn how to install the profile :-)

One more question: what are the relevant settings in firejail.config? So far I have only `network yes`. Should I also set `restricted-network yes`? Or should I let the latter commented out, which, I think, is the default option?

**Reply by ghost** [reply]
> Now I have to learn how to install the profile :-)

After downloading the files, move `disable-programs.local` and `ssmtp.profile` into ~/.config/firejail (applies to your user only) or in /etc/firejail (applies to all users). Make sure to chmod +x the `ssmtp` shell script and drop it in /usr/local/bin. Create the symlinks or just use `firejail /usr/bin/ssmtp <foo>` from CLI to test it.

> One more question: what are the relevant settings in firejail.config? So far I have only network yes. Should I also set restricted-network yes? Or should I let the latter commented out, which, I think, is the default option?

It should work without making any changes to firejail.config.

**Reply by Aqua1ung** [reply]
> After downloading the files, move `disable-programs.local` and `ssmtp.profile` into ~/.config/firejail (applies to your user only) or in /etc/firejail (applies to all users). Make sure to chmod +x the `ssmtp` shell script and drop it in /usr/local/bin. Create the symlinks or just use `firejail /usr/bin/ssmtp <foo>` from CLI to test it.

Thank you. Two questions:

1. shouldn't the CLI command be `firejail /usr/local/bin/ssmtp <foo>` instead of `firejail /usr/bin/ssmtp <foo>`?
2. wouldn't the `blacklist /etc/ssmtp` line in disable-programs.local prevent access to ssmtp.conf? I have crucial settings in ssmtp.conf, without which ssmtp won't work! Surely, ssmtp's settings need to be read from somewhere, and blacklisting ssmtp.conf would render it useless, woudn't it? [Edit: now I see a `noblacklist /etc/ssmtp` line in ssmtp.profile, so I am not sure I understand what's going on here :-) , but I'll just run with it and report back.]

**Reply by rusty-snake** [reply]
1. No, this would mean you start firejail in firejail which complicates things.
2. As you found the noblacklist you know that ssmtp is allowed to access `/etc/ssmtp`. However for every other firejailed process `/etc/ssmtp` is blacklisted.

**Reply by Aqua1ung** [reply]
> 1. No, this would mean you start firejail in firejail which complicates things.

There is no ssmtp file in /usr/bin!

**Reply by ghost** [reply]
> shouldn't the CLI command be firejail /usr/local/bin/ssmtp <foo> instead of firejail /usr/bin/ssmtp <foo>?

Good question. Let me try to explain. The /usr/local/bin/ssmtp script is a convenience wrapper. Because we place it in /usr/local/bin it should have precedence over the non-sandboxed ssmtp in /usr/bin. Just basic PATH logic at work here. With both scripts available, when you run `ssmtp foo`, what get's executed is the `firejail --quiet /usr/bin/ssmtp "$@"` part from /usr/local/bin/ssmtp. If you use `firejail /usr/local/bin/ssmtp` you'll end up with what we refer o as a 'doubled sandbox' because it comes down to executing `firejail firejail --quiet /usr/bin/ssmtp`. This will throw firejail into a fit, with unpredictable outcome and - as a matter of convention - is discouraged. I hope this clears up any confusion.

> wouldn't the blacklist /etc/ssmtp line in disable-programs.local prevent access to ssmtp.conf?

Yes and no. By itself this part of the files does indeed block access to ssmtp's configuration file. We do this to keep it out of reach of _other_ sandboxe that include disable-programs.inc. To override it so ssmtp can use it, we have the `noblacklist /etc/ssmtp` in ssmtp.profile. I do realize that it can take some time to get used by all these constructs. But after a while things start to click, come together and make sense.

Take your time to set things up correctly and experiment :-)

**Reply by ghost** [reply]
> There is no ssmtp file in /usr/bin!

You're probably using a Debian-like OS. Distributions package stuff differently and the provided `ssmtp` script uses /usr/sbin/ssmtp on systems like that. I do hope you have that file! The reason I wrote it like this is using two different Linux OS'es on the same machine and I want to support both in the same script. You could edit out this part and simply hardcode the correc path where your package manager installs the ssmtp binary. It's just an example. Once we've confirmed it works for you _and_ there's interest of adding it to our profiles collection, this sort of situation usually solves itself in the packaging stages/firecfg configuration. But it's always nice to have a clear understanding of the background anyway. Apologies for the potential confusion. The questions you pose indicate to me you'll do just fine.

**Reply by Aqua1ung** [reply]
> Good question. Let me try to explain. The /usr/local/bin/ssmtp script is a convenience wrapper. Because we place it in /usr/local/bin it should have precedence over the non-sandboxed ssmtp in /usr/bin.

Thanks. I placed the 3 files in the right spots, as instructed, and ran the following command (as root): `firejail --net=wlp1s0 /usr/bin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt` The result: `Error: line 70 in /etc/firejail/ssmtp.profile is invalid`. (The offending line is `restrict-namespaces`.)
What am I doing wrong? (Again, I do not have an ssmtp file in /usr/bin/!)

**Reply by ghost** [reply]
My bad. The `restrict-namespaces` is a recent addition. Your version doesn't have it yet and so the code throws an error and bails out. Just comment the line (put a # at the front) or remove it entirely.

We should stay on the same page, so I'll ask explicitly. Do you have /usr/sbin/ssmtp (mind the `sbin` part) on your system? Or could you provide a link to your distributions firejail package so I can check this?

**Reply by Aqua1ung** [reply]
Yes, that is where the ssmtp binary is, in /usr/sbin/ssmtp. And yes, running the `firejail --net=wlp1s0 /usr/bin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt` command now fails because it can't find some file, which I am guessing, is /usr/bin/ssmtp.

**Reply by ghost** [reply]
> Yes, that is where the ssmtp binary is, in /usr/sbin/ssmtp. And yes, running the firejail --net=wlp1s0 /usr/bin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt command now fails because it can't find some file, which I am guessing, is /usr/bin/ssmtp.

If like you say the ssmtp binary on your system is in /usr/sbin/ssmtp, you need to use that path in the command. Instead of `firejail --net=wlp1s0 /usr/bin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt` you should run `firejail --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt`. Could be a typo and something else goes wrong, but I cannot judge that from here. Here's a command that will produce a log file at /home/dad/ssmtp.log. Please paste its content here so we can have eyes on what's happening exactly.

```console
$ firejail --ignore=quiet --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt | tee /home/dad/ssmtp.log
```

**Reply by Aqua1ung** [reply]
Thank you for your attention to this matter, and I very much appreciate your bearing with me. I think it is time for a reset, if you do not mind: I think I am getting confused because you are, understandably, attempting to achieve more than what I am asking for. In particular, I am not interested in having firejail sandboxed _automatically_, nor am I interested in having sendmail sandboxed _automatically_: As such I have _refrained_ form adding any of the symlinks you've recommeded: none of these niceties are of interest to me, and I do not want to complicate an already complicated picture with any of these. (That being said, please let me know asap if the symlinks are, indeed, required for reasons _other_ than achieving desktop integration.)

So let me start from scratch, and recount the steps I take:

1. running Ubuntu KK (22.10);
2. install ssmtp; binary gets placed in `/usr/sbin/`;
3. install firejail: `sudo apt install firejail`;
4. switch `network yes` in `firejail.config`;
5. place _unmodified_ `disable-programs.local` and `ssmtp.profile` into `/etc/firejail`;
6. comment out line 70 in `ssmtp.profile`, no other changes made to `ssmtp.profile`;
7. place _unmodified_ `ssmtp` script into `/usr/local/bin`; make the `ssmtp` script executable;
8. run `sudo firejail --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt`;
9. this command, at this point in time, returns `command not found`;
10. run `$ firejail --ignore=quiet --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt | tee /home/dad/ssmtp.log`;
11. This also returns `command not found`.

So please let me know what I am doing wrong. Again, apologies for being so persistent (bordering on annoying)!

**Reply by rusty-snake** [reply]
https://github.com/netblue30/firejail/blob/65c92b464ef55b40a1005ad5b1c2932290c9dfd6/etc/inc/disable-common.inc#L470-L473

I guess so try to add

```
noblacklist /sbin
noblacklist /usr/local/sbin 
noblacklist /usr/sbin
```

**Reply by Aqua1ung** [reply]
@rusty-snake: you mean, edit these lines in `disable-common.inc` from `blacklist` to `noblacklist`?

**Reply by ghost** [reply]
> @rusty-snake: you mean, edit these lines in disable-common.inc from blacklist to noblacklist?

Correct. But do that in your `disable-common.local` instead. That overrides what's in /etc/firejail/disable-common.inc and is persistent. Anything you change in /etc/firejail/*.{inc,profile} will get wiped out on a future update of the firejail package.

**Reply by rusty-snake** [reply]
Even better in `ssmtp.profile`or`ssmtp.local`.

**Reply by Aqua1ung** [reply]
OK, added lines to ssmtp.profile. Here's what running `sudo firejail --ignore=quiet --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/fille.txt | tee /home/dad/ssmtp.log` returns:

```
tee: /home/dad/ssmtp.log: Permission denied
Reading profile /etc/firejail/ssmtp.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-programs.local
Reading profile /etc/firejail/disable-shell.inc
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 7718, child pid 7719

Interface        MAC                IP               Mask             Status
lo                                  127.0.0.1        255.0.0.0        UP    
eth0-7718        xx:xx:xx:xx:xx:xx  192.168.xxx.xx   255.255.255.0    UP    

6 programs installed in 11.48 ms
***
*** Warning: cannot whitelist ${DOCUMENTS} directory
*** Any file saved in this directory will be lost when the sandbox is closed.
***
***
*** Warning: cannot whitelist ${DOWNLOADS} directory
*** Any file saved in this directory will be lost when the sandbox is closed.
***
Blacklist violations are logged to syslog
Child process initialized in 1349.32 ms
Cannot start application: No such file or directory
The new log directory is /proc/7719/root/var/log

Parent is shutting down, bye...
```
And `ssmtp.log` is empty.

**Reply by ghost** [reply]
> *** Warning: cannot whitelist ${DOCUMENTS} directory
[...]
*** Warning: cannot whitelist ${DOWNLOADS} directory

Why do you keep running this with `sudo`? Obviously /root doesn't have a ${DOCUMENTS} nor a ${DOWNLOADS} folder. At least it shouldn't have on any sane system.

> tee: /home/dad/ssmtp.log: Permission denied
And ssmtp.log is empty.

Same remark. Try running without `sudo`!

**Reply by Aqua1ung** [reply]
@glitsj16: here's w/o sudo:

```
dad@mificommander:~/Downloads$ firejail --ignore=quiet --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/file.txt | tee /home/dad/ssmtp.log
tee: /home/dad/ssmtp.log: Permission denied
Reading profile /etc/firejail/ssmtp.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-programs.local
Reading profile /etc/firejail/disable-shell.inc
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 8119, child pid 8121

Interface        MAC                IP               Mask             Status
lo                                  127.0.0.1        255.0.0.0        UP    
eth0-8119        xx:xx:xx:xx:xx:xx  192.168.xxx.xx   255.255.255.0    UP    

6 programs installed in 11.47 ms
Blacklist violations are logged to syslog
Child process initialized in 1382.81 ms
Cannot start application: No such file or directory

Parent is shutting down, bye...
```

**Reply by ghost** [reply]
Argh, I forgot that `/home/dad` is read-only in the sandbox. Let's create `ssmtp.log` in a writeable location instead:

```console
$ firejail --ignore=quiet --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/file.txt | tee /home/dad/Downloads/ssmtp.log 2>&1
```

**Reply by Aqua1ung** [reply]
ssmtp.log in Downloads is still empty.

Here's what the command returns:

```
dad@mificommander:~$ firejail --ignore=quiet --net=wlp1s0 /usr/sbin/ssmtp -v my@email.com < /home/dad/Documents/file.txt | tee /home/dad/Downloads/ssmtp.log 2>&1
Reading profile /etc/firejail/ssmtp.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-programs.local
Reading profile /etc/firejail/disable-shell.inc
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Parent pid 8746, child pid 8748

Interface        MAC                IP               Mask             Status
lo                                  127.0.0.1        255.0.0.0        UP    
eth0-8746        xx:xx:xx:xx:xx:xx  192.168.xxx.xx   255.255.255.0    UP    

6 programs installed in 11.50 ms
Blacklist violations are logged to syslog
Child process initialized in 1385.48 ms
Cannot start application: No such file or directory

Parent is shutting down, bye...
```

**Reply by ghost** [reply]
Cannot understand what's going on at your side really. I did some more research, especially on ssmtp's configuration and packaging differences. Turned out to be quite the rabbit hole. If you want to drop these attempts on getting your ssmtp bound to your wireless network interface, I'd fully understand and would recommend using something like BindToInterface as mentioned in https://github.com/netblue30/firejail/discussions/5525#discussioncomment-4443156.

I do have it working on my Arch Linux box with Firejail and I also quadruple-checked my Ubuntu installation. On the latter I used [this PPA](https://launchpad.net/~team-mayhem/+archive/ubuntu/ppa). There are a few changes to my [gist](https://gist.github.com/glitsj16/76e6c6039e5d6ac4234b69d4686304ea). If you're still interested I suggest you have a look at the latest revision.

Good luck!

**Reply by Aqua1ung** [reply]
> I do have it working on my Arch Linux box with Firejail and I also quadruple-checked my Ubuntu installation. On the latter I used [this PPA](https://launchpad.net/~team-mayhem/+archive/ubuntu/ppa). There are a few changes to my [gist](https://gist.github.com/glitsj16/76e6c6039e5d6ac4234b69d4686304ea). If you're still interested I suggest you have a look at the latest revision.

Looks like Team Mayhem doesn't have a Kinetic repo. Can you point me to an ssmtp .deb file that you use for your own? Also, I am not wedded to ssmtp, so if you think I have better chances of success with other (preferably _very_ simple) MTAs, I'm willing to give it a try.

As for BindToInterface, that looks to be an even deeper  rabbit hole: tried to compile it as instructed, and I got a whole load of error messages: at this point, I don't even know if it's installed or not. Do you happen to know of a .deb file for that one?

**Reply by ghost** [reply]
> Team Mayhem doesn't have a Kinetic repo.

I picked the [Precise](https://launchpad.net/~team-mayhem/+archive/ubuntu/ppa/+files/ssmtp_2.65-0ppa1p_amd64.deb) version. But you can download any one of Team Mayhem's ssmtp .deb really. Dependencies shouldn't be a problem for Kinetic.

An alternative would be [msmtp](https://packages.ubuntu.com/search?suite=default&section=all&arch=any&searchon=names&keywords=msmtp). Also just a Mail Tranfer Agent. The Arch [wiki page](https://wiki.archlinux.org/title/Msmtp) on it has all the details.

> As for BindToInterface, that looks to be an even deeper rabbit hole: tried to compile it as instructed, and I got a whole load of error messages: at this point, I don't even know if it's installed or not. Do you happen to know of a .deb file for that one?

No .deb file for it that I'm aware of. Noticed your issue on it so I assume you got it compiled by now. Just `strip bindToInterface.so`, chmod 0755 and drop it in /usr/local/lib (or /usr/lib for that matter). Than you can bind ssmtp to your wlp1s0 via LD_PRELOAD:
`BIND_INTERFACE="wlp1s0" BIND_EXCLUDE="127.0.0.1" LD_PRELOAD="/usr/local/lib/bindToInterface.so" ssmtp ...`

**Reply by Aqua1ung** [reply]
Thanks! I just compiled the .c file, and moved the .so file thus obtained to `/usr/sbin`, and then did the `chmod 0755`. Is that good enough? What does `strip` do? Is that really needed?

**Reply by ghost** [reply]
Great. But these .so files are so called libraries or object files, not binaries, and you should follow convention here. /usr/sbin is not the location to put those in, move it to `/usr/lib` (which is what a .deb would do if there was one). The `strip` command removes unnneeded (debug) symbols and other data from object files. Same applies here, not srictly needed but what would happen if you install something packaged for Ubuntu via apt.

**Reply by ghost** [reply]
> Child process initialized in 1225.96 ms ssmtp: Cannot open smtp.comcast.net:465

Aha! I finally figured out what causes these errors. Doesn't make much sense, but if you add your user to the `mail` group things start to come together and ssmtp happily processes the message to the configured mailhub. As usual, relogin after adding your user to get going:

`sudo gpasswd --add <username> mail`

Maybe now I understand why you were running with `sudo` all this time. But there's really no need for that any longer.

**Reply by Aqua1ung** [reply]
Thanks. I did all that. As far as I can tell, it is still not working (see below): most of the time I get the same IP result with or without BIND_INTERFACE. (172.xxx.xxx.xxx is the IP of the (router connected via) wired NIC, and 98.xxx.xxx.xxx is the IP of the (router connected via) wireless NIC.) Pretty random. Interestingly, if I take a 5 mins break from launching this command, I get the right IP the first time, followed by the wrong IP afterwards.

```
dad@mificommander:~$ BIND_INTERFACE=wlp1s0 BIND_EXCLUDE=127.0.0.1 LD_PRELOAD=/usr/lib/bindToInterface.so curl ifconfig.me
98.xxx.xxx.xxx
dad@mificommander:~$ curl ifconfig.me
172.xxx.xxx.xxx
dad@mificommander:~$ BIND_INTERFACE=wlp1s0 BIND_EXCLUDE=127.0.0.1 LD_PRELOAD=/usr/lib/bindToInterface.so curl ifconfig.me
172.xxx.xxx.xxx
```

**Reply by Aqua1ung** [reply]
> > Child process initialized in 1225.96 ms ssmtp: Cannot open smtp.comcast.net:465
> 
> Aha! I finally figured out what causes these errors. Doesn't make much sense, but if you add your user to the `mail` group things start to come together and ssmtp happily processes the message to the configured mailhub. As usual, relogin after adding your user to get going:
> 
> `sudo gpasswd --add <username> mail`
> 
> Maybe now I understand why you were running with `sudo` all this time. But there's really no need for that any longer.

Does that mean that I can now run firejail with `--noprofile` and all that complicated jazz?

**Reply by ghost** [reply]
> Does that mean that I can now run firejail with --noprofile and all that complicated jazz?

That might work now, can't really tell. Just give it a try. We've already mentioned that --noprofile doesn't offer any security, but you said you don't want that anyway. That's entirely up to you.

**Reply by ghost** [reply]
> Thanks. I did all that. As far as I can tell, it is still not working (see below): most of the time I get the same IP result with or without BIND_INTERFACE. (172.xxx.xxx.xxx is the IP of the (router connected via) wired NIC, and 98.xxx.xxx.xxx is the IP of the (router connected via) wireless NIC.) Pretty random. Interestingly, if I take a 5 mins break from launching this command, I get the right IP the first time, followed by the wrong IP afterwards.

Sounds like your routing tables or firewall rules are causing this. Or a VPN. Out-of-scope for me though, I'm not very knowledgeable in that area.

**Comment by ghost** [comment]
> I need to bind ssmtp to my wireless network interface. Can this be done using firejail?

It can, as per https://github.com/netblue30/firejail/discussions/5525#discussioncomment-4403289.

> Thank you for your attention to this matter, and I very much appreciate your bearing with me. I think it is time for a reset, if you do not mind: I think I am getting confused because you are, understandably, attempting to achieve more than what I am asking for. In particular, I am not interested in having firejail sandboxed automatically, nor am I interested in having sendmail sandboxed automatically.

No problem, I do understand we added complexity. For a non-sandboxed, easy solution I suggest you use something like [BindToInterface](https://github.com/JsBergbau/BindToInterface).
